File: ReactNative/api/colonias.js
import axios from 'axios';
import API_CONFIG from './config';

/**
 * API para gestión de colonias
 */
const ColoniasAPI = {
  /**
   * Crea una nueva colonia
   * @param {Object} data Datos de la colonia
   * @param {string} token Token de autenticación (opcional)
   * @returns {Promise<Object>} Respuesta de la API
   */
  createColonia: async (data, token = null) => {
    try {
      // Configurar headers
      const headers = token 
        ? { ...API_CONFIG.HEADERS, ...API_CONFIG.getAuthHeader(token) }
        : API_CONFIG.HEADERS;
      
      // Realizar petición
      API_CONFIG.logRequest('POST', `${API_CONFIG.BASE_URL}/colonias/`, data);
      
      const response = await axios.post(
        `${API_CONFIG.BASE_URL}/colonias/`, 
        data,
        { headers }
      );
      
      API_CONFIG.logResponse(response);
      return { success: true, data: response.data };
    } catch (error) {
      API_CONFIG.logError(error);
      return { 
        success: false, 
        error: error.response?.data?.detail || error.message,
        status: error.response?.status
      };
    }
  },

  /**
   * Obtiene todas las colonias
   * @param {string} token Token de autenticación (opcional)
   * @returns {Promise<Object>} Respuesta de la API
   */
  getColonias: async (token = null) => {
    try {
      const headers = token 
        ? { ...API_CONFIG.HEADERS, ...API_CONFIG.getAuthHeader(token) }
        : API_CONFIG.HEADERS;
      
      API_CONFIG.logRequest('GET', `${API_CONFIG.BASE_URL}/colonias/`);
      
      const response = await axios.get(
        `${API_CONFIG.BASE_URL}/colonias/`,
        { headers }
      );
      
      API_CONFIG.logResponse(response);
      return { success: true, data: response.data };
    } catch (error) {
      API_CONFIG.logError(error);
      return { 
        success: false, 
        error: error.response?.data?.detail || error.message,
        status: error.response?.status
      };
    }
  },

  /**
   * Obtiene una colonia por su ID
   * @param {string} id ID de la colonia
   * @param {string} token Token de autenticación (opcional)
   * @returns {Promise<Object>} Respuesta de la API
   */
  getColoniaById: async (id, token = null) => {
    try {
      const headers = token 
        ? { ...API_CONFIG.HEADERS, ...API_CONFIG.getAuthHeader(token) }
        : API_CONFIG.HEADERS;
      
      API_CONFIG.logRequest('GET', `${API_CONFIG.BASE_URL}/colonias/${id}`);
      
      const response = await axios.get(
        `${API_CONFIG.BASE_URL}/colonias/${id}`,
        { headers }
      );
      
      API_CONFIG.logResponse(response);
      return { success: true, data: response.data };
    } catch (error) {
      API_CONFIG.logError(error);
      return { 
        success: false, 
        error: error.response?.data?.detail || error.message,
        status: error.response?.status
      };
    }
  }
};

export default ColoniasAPI;

File: FastAPI/routes/colonias.py
from fastapi import APIRouter, Depends, HTTPException
from typing import List
import logging
from uuid import UUID
from database import get_db
from models.colonia import Colonia, ColoniaCreate
from models.residencia import Residencia
from crud.colonias import create_colonia, get_colonias, get_colonia_by_id
from crud.residencias import get_residencias_by_colonia
from crud.base import serialize_model

router = APIRouter()
logger = logging.getLogger(__name__)

@router.post("/", response_model=Colonia)
async def create_colonia_route(colonia_data: ColoniaCreate, db=Depends(get_db)):
    """Crea una nueva colonia."""
    try:
        # Crear un objeto Colonia completo a partir de ColoniaCreate
        colonia = Colonia(
            nombre=colonia_data.nombre,
            direccion=colonia_data.direccion,
            admin_principal_id=colonia_data.admin_principal_id
            # Los demás campos se generan automáticamente
        )
        created_colonia = create_colonia(colonia, db)
        return created_colonia
    except Exception as e:
        # Logging detallado para depuración
        logger.error(f"Error al crear colonia: {str(e)}")
        if isinstance(e, HTTPException):
            raise e
        raise HTTPException(status_code=500, detail=f"Error interno: {str(e)}")

@router.get("/", response_model=List[Colonia])
async def get_colonias_route(db=Depends(get_db)):
    colonias = get_colonias(db)
    return colonias

@router.get("/{colonia_id}/residencias", response_model=List[Residencia])
async def get_residencias_by_colonia_route(colonia_id: UUID, db=Depends(get_db)):
    """Obtiene todas las residencias de una colonia."""
    residencias = get_residencias_by_colonia(colonia_id, db)
    return residencias

@router.get("/{colonia_id}/generar-codigo/{cantidad}")
async def generar_codigos_residencias_route(colonia_id: UUID, cantidad: int, db=Depends(get_db)):
    """Genera un número específico de códigos para residencias en una colonia."""
    # Verificar que la colonia existe
    colonia_response = db.table('colonias').select('*').eq('id', str(colonia_id)).execute()
    if not colonia_response.data:
        raise HTTPException(status_code=404, detail="Colonia no encontrada")
    
    residencias_generadas = []
    
    # Generar las residencias con sus códigos
    for i in range(cantidad):
        nueva_residencia = Residencia(
            numero=f"Auto-{i+1}",  # Número provisional
            calle="Por asignar",   # Calle provisional
            referencia="Residencia generada automáticamente",
            colonia_id=colonia_id
        )
        
        # Serializar y guardar
        serialized_data = serialize_model(nueva_residencia)
        response = db.table('residencias').insert(serialized_data).execute()
        
        if response.data:
            # Guardar la información con el ID generado
            residencia_guardada = response.data[0]
            residencias_generadas.append({
                "id": residencia_guardada["id"],
                "codigo": residencia_guardada["id"]  # El UUID como código
            })
    
    return {"residencias_generadas": residencias_generadas}